src/core/symbology-ng/qgsgeometrygeneratorsymbollayerv2.cpp

```diff
@@ -186,17 +186,22 @@ bool QgsGeometryGeneratorSymbolLayerV2::isCompatibleWithSymbol( QgsSymbolV2* sym
   Q_UNUSED( symbol )
   return true;
 }
-
 void QgsGeometryGeneratorSymbolLayerV2::render( QgsSymbolV2RenderContext& context )
 {
-  QgsGeometry geom = mExpression->evaluate( &context.renderContext().expressionContext() ).value<QgsGeometry>();
   if ( context.feature() )
   {
-    QgsFeature f = *context.feature();
+    QgsExpressionContext& expressionContext = context.renderContext().expressionContext();
+
+    QgsFeature f = expressionContext.feature();
```

Comment:
The previous version did not replace the geometry in the expression scope but only passed it to the symbol render function so there may have been unwanted side-effects.

Now we leave the original feature untouched (in a less-important scope) and add a new scope with a modified symbol version that hides the other one. This seems much cleaner.